chapter_recurrent-modern/gru.ipynb

```diff
@@ -0,0 +1,482 @@
+    "    NDArray b_q = manager.zeros(new Shape(numOutputs), DataType.FLOAT32, device);\n",
+    "\n",
+    "    // Attach gradients\n",
+    "    NDList params = new NDList(W_xz, W_hz, b_z, W_xr, W_hr, b_r, W_xh, W_hh, b_h, W_hq, b_q);\n",
+    "    for (NDArray param : params) {\n",
+    "        param.setRequiresGradient(true);\n",
+    "    }\n",
+    "    return params;\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 6
+   },
+   "source": [
+    "### Defining the Model\n",
+    "\n",
+    "Now we will define the hidden state initialization function `init_gru_state`. Just like the `init_rnn_state` function defined in :numref:`sec_rnn_scratch`, this function returns a tensor with a shape (batch size, number of hidden units) whose values are all zeros.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public static NDList initGruState(int batchSize, int numHiddens, Device device) {\n",
+    "    return new NDList(manager.zeros(new Shape(batchSize, numHiddens), DataType.FLOAT32, device));\n",
+    "}"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "origin_pos": 9
+   },
+   "source": [
+    "Now we are ready to define the GRU model.\n",
+    "Its structure is the same as that of the basic RNN cell, except that the update equations are more complex.\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "public static Pair<NDArray, NDList> gru(NDArray inputs, NDList state, NDList params) {\n",
+    "    NDArray W_xz = params.get(0);\n",
+    "    NDArray W_hz = params.get(1);\n",
+    "    NDArray b_z = params.get(2);\n",
+    "\n",
+    "    NDArray W_xr = params.get(3);\n",
+    "    NDArray W_hr = params.get(4);\n",
+    "    NDArray b_r = params.get(5);\n",
+    "\n",
+    "    NDArray W_xh = params.get(6);\n",
+    "    NDArray W_hh = params.get(7);\n",
+    "    NDArray b_h = params.get(8);\n",
+    "\n",
+    "    NDArray W_hq = params.get(9);\n",
+    "    NDArray b_q  = params.get(10);\n",
+    "\n",
+    "    NDArray H = state.get(0);\n",
+    "    NDList outputs = new NDList();\n",
+    "    NDArray X, Y, Z, R, H_tilda;\n",
+    "    for (int i = 0; i < inputs.size(0); i++) {\n",
+    "        X = inputs.get(i);\n",
+    "        Z = Activation.sigmoid(X.dot(W_xz).add(H.dot(W_hz).add(b_z)));\n",
+    "        R = Activation.sigmoid(X.dot(W_xr).add(H.dot(W_hr).add(b_r)));\n",
+    "        H_tilda = Activation.tanh(X.dot(W_xh).add(R.mul(H).dot(W_hh).add(b_h)));\n",
+    "        H = Z.mul(H).add(Z.mul(-1).add(1).mul(H_tilda));\n",
+    "        Y = H.dot(W_hq).add(b_q);\n",
+    "        outputs.add(Y);\n",
+    "    }\n",
+    "    return new Pair(outputs.size() > 1 ? NDArrays.concat(outputs) : outputs.get(0), new NDList(H));\n",
```

Comment:
Doesn't it work on outputs.size() == 1?

Reply:
Nope, if I am not mistaken that's the reason I had done that. It threw an error otherwise.